cpp/src/quantiles/quantiles.cu

```diff
@@ -1,109 +1,90 @@
-/*
- * Copyright (c) 2019, NVIDIA CORPORATION.
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
 #include <memory>
-#include <thrust/functional.h>
-#include <thrust/iterator/counting_iterator.h>
-#include <cudf/copying.hpp>
-#include <cudf/scalar/scalar.hpp>
-#include <cudf/table/row_operators.cuh>
-#include <cudf/table/table_device_view.cuh>
+#include <vector>
+
 #include <cudf/types.hpp>
-#include <cudf/utilities/traits.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/detail/gather.cuh>
 #include <quantiles/quantiles_util.hpp>
-#include "cudf/utilities/error.hpp"
 
 namespace cudf {
 namespace experimental {
-namespace detail {
-namespace {
 
-using ScalarResult = double;
+namespace detail {
 
-struct quantile_functor
+template<typename SortMapIterator>
+std::unique_ptr<table>
+quantiles(table_view const& input,
+          SortMapIterator sortmap,
+          std::vector<double> const& q,
+          interpolation interp,
+          rmm::mr::device_memory_resource* mr)
 {
-    template<typename T, typename... Args>
-    std::enable_if_t<not std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
-    operator()(Args&&... args)
-    {
-        CUDF_FAIL("Only numeric types are supported in quantiles.");
-    }
-
-    template<typename T>
-    std::enable_if_t<std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
-    operator()(column_view const& input,
-               double percent,
-               interpolation interp,
-               order_info column_order,
-               rmm::mr::device_memory_resource *mr =
-                 rmm::mr::get_default_resource(),
-               cudaStream_t stream = 0)
-    {
-        if (input.size() == input.null_count()) {
-            return std::make_unique<numeric_scalar<ScalarResult>>(0, false, stream);
-        }
+    auto sortmap_lookup = [sortmap]
```

Comment:
What's the purpose of putting this in a separate lambda instead of just putting the same logic inside the `quantile_idx_lookup` lambda?

Reply:
This is just an artifact of me playing around with iterators instead of a selector, and it got left in an intermediary state.

Reply:
Ah, I may have misinterpreted your question.

If you mean why do two lambdas need to exist, it's because I made a mistake when implementing `select_quantile`. It uses a function rather than an iterator, so instead of passing the sortmap directly, it has to first be wrapped in a function.

If you mean why is the first lamba not nested within the first, the answer is there is no reason.